builq.go

```diff
@@ -5,57 +5,50 @@ import (
 	"strings"
 )
 
-// Builder for SQL queries.
+// TODO(junk1tm): support PostgreSQL-style numbered placeholders.
+var Placeholder = "$"
+
+// Builder for SQL queries. The zero value is ready to use, just like strings.Builder.
 type Builder struct {
 	query strings.Builder
 	args  []interface{}
 }
 
-// Newf creates a new query
-func Newf(format string, args ...interface{}) *Builder {
-	b := &Builder{}
-	q := fmt.Sprintf(format, args...)
-	b.query.WriteString(q)
-	return b
-}
-
-// Query that was build.
-func (b *Builder) Query() string { return b.query.String() }
-
-// Args passed to the query.
-func (b *Builder) Args() []interface{} { return b.args }
-
-// Append the expression with prepended "\n"
-// Optional expressions are prepended with " ".
-func (b *Builder) Append(expr string, exprs ...string) {
-	b.query.WriteString("\n" + expr)
-	for _, expr := range exprs {
-		b.query.WriteByte(' ')
-		b.query.WriteString(expr)
+func (b *Builder) Appendf(format string, args ...interface{}) *Builder {
+	var wargs []interface{}
+	for _, arg := range args {
+		wargs = append(wargs, &argument{value: arg})
+	}
+	fmt.Fprintf(&b.query, format, wargs...)
+	for _, warg := range wargs {
+		if arg := warg.(*argument); arg.forQuery {
+			b.args = append(b.args, arg.value)
+		}
 	}
+	return b
 }
 
-// Add the expression and the parameter and return its index in query.
-func (b *Builder) Add(expr string, v interface{}) string {
-	param := b.AddParam(v)
-	b.query.WriteString("\n" + expr + param)
-	return param
+func (b *Builder) Build() (string, []interface{}, error) {
+	return b.query.String(), b.args, nil
 }
 
-// AddParam add parameter and return its index in query.
-func (b *Builder) AddParam(v interface{}) string {
-	b.args = append(b.args, v)
-	return fmt.Sprintf("$%d", len(b.args))
+// argument is a wrapper for Printf-style arguments that implements fmt.Formatter.
+type argument struct {
+	value    interface{}
+	forQuery bool // is it a query argument?
 }
 
-// AddParams add parameters and return its index in query.
-func (b *Builder) AddParams(v ...interface{}) string {
-	res := ""
-	for i, v := range v {
-		if i > 0 {
-			res += ", "
-		}
-		res += b.AddParam(v)
+// Format implements the fmt.Formatter interface.
+func (a *argument) Format(s fmt.State, v rune) {
+	switch v {
+	case 's':
+		// just a normal string (a table, a column, etc.), write it as is.
+		fmt.Fprint(s, a.value)
+	case 'a':
+		// a query argument, mark it and write a placeholder.
+		a.forQuery = true
+		fmt.Fprint(s, Placeholder)
+	default:
+		panic(fmt.Sprintf("unsupported verb %c", v))
```

Comment:
I'm not sure that panic is ok here. I think we should do the same as fmt - catch panic, print error to `s` (even if it's `fmt.State`) and also return error in `Build()` method (where we have `nil` now).

This should make behaviour more predictive + make error in `Build()` useful. WDYT ?

Also, this can be done in a separate PR.

Reply:
100% agree, will be done in a separate PR :)